libmambapy/src/libmambapy/bindings/specs.cpp

```diff
@@ -0,0 +1,466 @@
+                },
+                py::arg("decode") = true
+            )
+            .def(
+                "set_package",
+                [](CondaURL& self, std::string_view str, bool encode)
+                {
+                    if (encode)
+                    {
+                        return self.set_package(str, CondaURL::Encode::yes);
+                    }
+                    return self.set_package(std::string(str), CondaURL::Encode::no);
+                },
+                py::arg("package"),
+                py::arg("encode") = true
+            )
+            .def("clear_package", &CondaURL::clear_package)
+            .def("str", &CondaURL::str, py::arg("credentials") = CondaURL::Credentials::Hide)
+            .def(
+                "__repr__",
+                [](const CondaURL& self) { return self.str(CondaURL::Credentials::Hide); }
+            )
+            .def(
+                "pretty_str",
+                [](const CondaURL& self,
+                   bool strip_scheme,
+                   std::string_view rstrip_path,
+                   CondaURL::Credentials credentials)
+                {
+                    auto const scheme = strip_scheme ? CondaURL::StripScheme::yes
+                                                     : CondaURL::StripScheme::no;
+                    const char rstrip = rstrip_path.empty() ? '\0' : rstrip_path.front();
+                    return self.pretty_str(scheme, rstrip, credentials);
+                },
+                py::arg("strip_scheme") = false,
+                py::arg("rstrip_path") = "/",
+                py::arg("credentials") = CondaURL::Credentials::Hide
+            );
+
+        auto py_channel_spec = py::class_<ChannelSpec>(m, "ChannelSpec");
+
+        py::enum_<ChannelSpec::Type>(py_channel_spec, "Type")
+            .value("URL", ChannelSpec::Type::URL)
+            .value("PackageURL", ChannelSpec::Type::PackageURL)
+            .value("Path", ChannelSpec::Type::Path)
+            .value("PackagePath", ChannelSpec::Type::PackagePath)
+            .value("Name", ChannelSpec::Type::Name)
+            .value("Unknown", ChannelSpec::Type::Unknown)
+            .def(py::init(&enum_from_str<ChannelSpec::Type>));
+        py::implicitly_convertible<py::str, ChannelSpec::Type>();
+
+        py_channel_spec  //
+            .def_static("parse", ChannelSpec::parse)
+            .def(
+                py::init<std::string, ChannelSpec::dynamic_platform_set, ChannelSpec::Type>(),
+                py::arg("location"),
+                py::arg("platform_filters"),
+                py::arg("type") = ChannelSpec::Type::Unknown
+            )
+            .def("__copy__", &copy<ChannelSpec>)
+            .def("__deepcopy__", &deepcopy<ChannelSpec>, pybind11::arg("memo"))
+            .def_property("type", &ChannelSpec::type, &ChannelSpec::set_type)
+            .def_property("location", &ChannelSpec::location, &ChannelSpec::set_location)
```

Comment:
I guess these are the culprits for the setters in the C++ code. Do we really need that on the Python side? Can't we require the objects to be constructed once and for good?